### PR TITLE
vam: divide by zero error

### PR DESCRIPTION
**Pull request by mattnibs**

### CODE REVIEW

File: runtime/vam/expr/arith.go [hunk 1]
```diff
@@ -48,5 +51,90 @@ func (a *Arith) eval(vecs ...vector.Any) vector.Any {
 	if !ok {
 		return vector.NewStringError(a.zctx, coerce.ErrIncompatibleTypes.Error(), lhs.Len())
 	}
-	return f(lhs, rhs)
+	out, ok := f(lhs, rhs)
+	if !ok {
+		return a.evalDivideByZero(kind, lhs, rhs)
+	}
+	return out
```
Comment (nwt) [inline runtime/vam/expr/arith.go:58]: Keep genarithfuncs.go a little simpler by letting the hardware do the check?
```suggestion
	if a.opCode == vector.ArithDiv || a.opCode == vector.ArithMod {
		defer func() {
			if v := recover(); v != nil {
				if err, ok := v.(runtime.Error); ok && err.Error() == "runtime error: integer divide by zero" {
					out = a.evalDivideByZero(kind, lhs, rhs)
				} else {
					panic(v)
				}
			}
		}()
	}
	return f(lhs, rhs)	
```

File: runtime/vam/expr/arith.go [hunk 2]
```diff
@@ -48,5 +51,90 @@ func (a *Arith) eval(vecs ...vector.Any) vector.Any {
...
+				floats = append(floats, 0)
+				continue
+			}
+			if r == 0 {
+				errs = append(errs, i)
+				continue
+			}
+			floats = append(floats, l/r)
+		}
+		out = vector.NewFloat(super.TypeFloat64, floats, nulls)
```
Comment (nwt) [inline runtime/vam/expr/arith.go:132]: Floating point division by zero isn't usually an error. (The IEEE 754 spec says it results in NaN, +Inf, or -Inf depending on the operands.)
```suggestion
```

File: runtime/vam/expr/arith.go (hunk 1 above)
Comment (mattnibs) [reply runtime/vam/expr/arith.go:58]: Yeah I was thinking about this sounds good

File: runtime/vam/expr/arith.go [hunk 3]
```diff
@@ -48,5 +52,74 @@ func (a *Arith) eval(vecs ...vector.Any) vector.Any {
...
+				if err, ok := v.(runtime.Error); ok && err.Error() == "runtime error: integer divide by zero" {
+					out = a.evalDivideByZero(kind, lhs, rhs)
+				} else {
+					panic(v)
+				}
+			}
+		}()
+	}
+	out = f(lhs, rhs)
+	return
```
Comment (nwt) [inline runtime/vam/expr/arith.go:67]: Nit: Naked returns are the worst.
```suggestion
	return f(lhs, rhs)
```

File: runtime/vam/expr/genarithfuncs.go [hunk 4]
```diff
@@ -49,7 +49,7 @@ func main() {
 		}
 	}
 
-	fmt.Fprintln(&buf, "var arithFuncs = map[int]func(vector.Any, vector.Any) vector.Any{")
+	fmt.Fprintln(&buf, "var arithFuncs = map[int]func(vector.Any, vector.Any) vector.Any {")
```
Comment (nwt) [inline runtime/vam/expr/genarithfuncs.go:52]: Nit: Don't change this file.

**Review by nwt — APPROVED** [review]

**Pull request merged**